### PR TITLE
rats-tls: support sev(-es) remote attestation

### PR DESCRIPTION
**Pull request by haosanzi**

### CODE REVIEW

File: src/verifiers/sev/amdcert.c [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+		return -1;
+	}
+
+	memset(pub_key_cert, 0, sizeof(*pub_key_cert));
+
+	if (cert->modulus_size == AMD_CERT_KEY_BITS_2K) {
+		pub_key_cert->pub_key_algo = SEV_SIG_ALGO_RSA_SHA256;
+	} else if (cert->modulus_size == AMD_CERT_KEY_BITS_4K) {
+		pub_key_cert->pub_key_algo = SEV_SIG_ALGO_RSA_SHA384;
+	}
```
Comment (jiazhang0) [inline src/verifiers/sev/amdcert.c:226]: plz add a assert() declaration here.

File: src/verifiers/sev/amdcert.c [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+			return false;
+	}
+
+	const uint8_t *amd_root_key_id = NULL;
+	if (device_type == PSP_DEVICE_TYPE_NAPLES)
+		amd_root_key_id = amd_root_key_id_naples;
+	else if (device_type == PSP_DEVICE_TYPE_ROME)
+		amd_root_key_id = amd_root_key_id_rome;
+	else if (device_type == PSP_DEVICE_TYPE_MILAN)
+		amd_root_key_id = amd_root_key_id_milan;
```
Comment (jiazhang0) [inline src/verifiers/sev/amdcert.c:202]: plz add a assert() declaration here.

File: src/verifiers/sev-snp/sevapi.h [hunk 3]
```diff
@@ -16,6 +16,11 @@ typedef uint8_t hmac_sha_256[256 / 8];
 typedef uint8_t hmac_sha_384[384 / 8];
 typedef uint8_t hmac_sha_512[512 / 8];
 
+typedef enum __attribute__((mode(QI))) SHA_TYPE {
+	SHA_TYPE_256 = 0,
+	SHA_TYPE_384 = 1,
+} SHA_TYPE;
```
Comment (jiazhang0) [inline src/verifiers/sev-snp/sevapi.h:22]: Change to SHA_ALGO.

File: src/verifiers/sev/sevcert.c [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+		hash_len = sizeof(hmac_sha_256);
+		break;
+	case SEV_SIG_ALGO_RSA_SHA384:
+	case SEV_SIG_ALGO_ECDSA_SHA384:
+	case SEV_SIG_ALGO_ECDH_SHA384:
+		hash_algo = SHA_TYPE_384;
+		hash = sha384_digest;
+		hash_len = sizeof(hmac_sha_384);
+		break;
+	default:
```
Comment (jiazhang0) [inline src/verifiers/sev/sevcert.c:189]: add error print here

File: src/verifiers/sev/sevcert.c [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+		if (parent_cert1->pub_key_usage != SEV_USAGE_OCA)
+			goto err;
+		break;
+	case SEV_USAGE_CEK:
+		/* The CEK must be signed by the ASK */
+		if (parent_cert1->pub_key_usage != SEV_USAGE_ASK)
+			goto err;
+		break;
+	default:
+		goto err;
```
Comment (jiazhang0) [inline src/verifiers/sev/sevcert.c:362]: add error print here

File: src/verifiers/sev/verify_evidence.c [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+		ark_ask_cert_patch = SEV_ROME_DEFAULT_DIR ASK_ARK_FILENAME;
+		url = ASK_ARK_ROME_SITE;
+		break;
+	case PSP_DEVICE_TYPE_MILAN:
+		default_dir = SEV_MILAN_DEFAULT_DIR;
+		ark_ask_cert_patch = SEV_MILAN_DEFAULT_DIR ASK_ARK_FILENAME;
+		url = ASK_ARK_MILAN_SITE;
+		break;
+	default:
+		RTLS_ERR("unsupported device type\n");
```
Comment (jiazhang0) [inline src/verifiers/sev/verify_evidence.c:41]: print the value of unsupported device type.

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: src/attesters/sev/init.c [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+#include <rats-tls/attester.h>
+
+static unsigned int dummy_private;
+
+enclave_attester_err_t sev_attester_init(enclave_attester_ctx_t *ctx, rats_tls_cert_algo_t algo)
+{
+	RTLS_DEBUG("ctx %p, algo %d\n", ctx, algo);
+
+	ctx->attester_private = &dummy_private;
+
```
Comment (zhiminghufighting) [inline src/attesters/sev/init.c:17]: This static variable is initialized may be more better even it is not used in current scenario.

File: src/attesters/api/enclave_attester_register.c [hunk 8]
```diff
@@ -38,6 +38,16 @@ enclave_attester_err_t enclave_attester_register(const enclave_attester_opts_t *
...
+		if (!is_sevguest_supported()) {
+			// clang-format off
+			RTLS_DEBUG("failed to register the attester '%s' due to lack of SEV(-ES) Guest capability\n",
+				   opts->type);
+			// clang-format on
+			return -ENCLAVE_ATTESTER_ERR_CPU_UNSUPPORTED;
+		}
+	}
+
 	enclave_attester_opts_t *new_opts = (enclave_attester_opts_t *)malloc(sizeof(*new_opts));
```
Comment (zhiminghufighting) [inline src/attesters/api/enclave_attester_register.c:51]: Don't suggest to put self-detect function here. We can put it in its' init.c or pre_init.c.

Comment (haosanzi) [reply src/attesters/api/enclave_attester_register.c:51]: Good point! I suggest detecting the registered environment according to the previous logic and then modifying both sev, snp and tdx uniformly after the out-of-tree code is completed.

File: src/attesters/sev/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,40 @@
+# Project name
+project(attester_sev)
+
+# Set include directory
+set(INCLUDE_DIRS ${CMAKE_CURRENT_SOURCE_DIR}/../../include
+                 ${CMAKE_CURRENT_SOURCE_DIR}/../../include/enclave-tls
```
Comment (YangLiang3) [inline src/attesters/sev/CMakeLists.txt:6]: Are you sure enclave-tls directory is there?

File: src/verifiers/sev/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,35 @@
+# Project name
+project(verifier_sev)
+
+# Set include directory
+set(INCLUDE_DIRS ${CMAKE_CURRENT_SOURCE_DIR}/../../include
+                 ${CMAKE_CURRENT_SOURCE_DIR}/../../include/enclave-tls
```
Comment (YangLiang3) [inline src/verifiers/sev/CMakeLists.txt:6]: please check the path

File: src/verifiers/sev/amdcert.c [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+#include <string.h>
+#include <openssl/evp.h>
+#include <openssl/hmac.h>
+#include <openssl/rsa.h>
+#include <openssl/sha.h>
+#include "amdcert.h"
+#include "../sev-snp/utils.h"
+#include "../sev-snp/crypto.c"
+
+/* Bytes, NOT bits */
```
Comment (YangLiang3) [inline src/verifiers/sev/amdcert.c:16]: please add more description details

**Review by jiazhang0 — APPROVED** [review]

**Review by YangLiang3 — APPROVED** [review]

**Review by zhiminghufighting — APPROVED** [review]

**Pull request merged**